rfcs/src/rfcs/0001-mir-linker.md

```diff
@@ -0,0 +1,244 @@
+
+### Alternatives
+
+The other options explored were:
+1. Pre-compile the standard library and ship the generated `*symtab.json` files.
+2. Pre-compile the standard library, convert them to `goto-programs` (via `symtab2gb`) and link them into
+   one single `goto-program` model. Ship the generated model.
+
+Both alternatives are very similar. They only differ on the artifact that would be shipped.
+They require generating and shipping a custom `sysroot`;
+however, there is no need to implement the reachability algorithm.
+
+We implemented a prototype for the MIR linker and one for the alternatives.
+Both prototypes generate the sysroot as part of the `cargo kani` flow.
+
+We performed a small experiment (on a `c5.4xlarge` ec2 instance running Ubuntu 20.04) to assess the options.
+
+For this experiment, we used the following harness:
+```rust
+#[kani::proof]
+#[kani::unwind(4)]
+pub fn check_format() {
+    assert!("2".parse::<u32>().unwrap() == 2);
+}
+```
+The experiment showed that the MIR linker approach is much more efficient.
+
+See the table bellow for the breakdown of time (in seconds) taken for each major step of
+the harness verification:
+
+
+| Stage                     | MIR Linker | Alternative |
+----------------------------|------------|-------------|
+| compilation               |   22.2s    |    64.7s    |
+| `goto-program` generation |   2.4s     |    90.7s    |
+| `goto-program` linking    |   0.8s     |    33.2s    |
+| code instrumentation      |   0.8s     |    33.1     |
+| verification              |   0.5s     |    8.5s     |
+
+It is possible that `goto-cc` time can be improved, but this would also require further experimentation and
+expertise that we don't have today.
+
+Every option would require a custom sysroot to either be built or shipped with Kani.
+The table below shows the size of the sysroot files for the alternative #2
+(`goto-program` files) vs compiler artifacts (`*.rmeta` files)
+files with `-Z always-encode-mir` for `x86_64-unknown-linux-gnu` (on Ubuntu 18.04).
+
+| File Type      | Raw size | Compressed size |
+|----------------|----------|-----------------|
+| `symtab.json`  |   950M   |     26M         |
+| `symtab.out`   |   84M    |     24M         |
+| `*.rmeta`      |   92M    |     25M         |
+
+These results were obtained by looking at the artifacts generated during the same experiment.
+
+## Open questions
+
+- Should we build or download the sysroot during `kani setup`?
+- What's the best way to enable support to run Kani in the entire `workspace`?
+  - One possibility is to run `cargo rustc` per package.
+- Should we codegen all static items no matter what? Static objects can only be initialized via constant function.
+  Thus, it shouldn't have any side effect.
+  That relies on all constant initializers being evaluated during compilation.
+
+## Future possibilities
+
+- Split `goto-program` models into two or more items to optimize compilation result caching.
```

Comment:
A related question: are we perhaps able to skip re-generating the MIR for a dependent crate that hasn't changed, and reuse it in the following steps? `cargo build` (and hence `cargo kani`) sort of already does that when compiling a package, but we always run `symtab2gb` even if the symtab file hasn't changed.

Reply:
We would be able to skip re-generating the MIR for the dependencies but we would still generate the goto-program to everything linked at every compilation.

Being able to cache the goto-program generation for things that haven't changed would require more work and this is what I'm trying to capture here.

Reply:
Splitting is one possibility, but IMO before we pursue any of that, we should take seriously optimizing the MIR -> Goto binary process. This would benefit initial as well as subsequent, and I think there's probably a LOT of low-hanging fruit.

Both in our compiler, symtab2gb, and the vague future possibility that we could emit goto binaries directly without going through json and symtab2gb.